[zulu-openjdk] Bugfixes for the Zulu JDK package

datalogics-robb opened the pull request:
Recently the Zulu package was enhanced by adding additional binaries for version 11.0.24.
However, other changes to support these binary packages should also have been made at that time, as none of them were functional.
This pull request resolves three of these problems and has been tested on these platforms.

### Summary
Changes to recipe:  **zulu-openjdk/11.0.24**

#### Motivation

host settings should identfy the binary type of the package:
closes https://github.com/conan-io/conan-center-index/issues/25484

validation should support all the new configurations:
closes https://github.com/conan-io/conan-center-index/issues/25483

Zip files with executables need an extra flag:
closes https://github.com/conan-io/conan-center-index/issues/25482

#### Details
 - Use host settings instead of build settings (see issue #25484 for details)
 - validate based on the contents of conandata.yml
 - use the keyword argument `keep_permissions=True` for the `get()` tool

---
- [x] Read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md)
- [x] Checked that this PR is not a duplicate: [list of PRs by recipe](https://github.com/conan-io/conan-center-index/discussions/24240)
- [x] Tested locally with at least one configuration using a recent version of Conan

datalogics-robb committed:
[zulu-openjdk] Bugfixes for the Zulu JDK package

Recently the Zulu package was enhance by
adding many different binaries for version 11.0.24.
However, other changes to support these binary packages should
also have been made at that time, as none of them were functional.

host settings should identfy the binary type of the package:
 https://github.com/conan-io/conan-center-index/issues/25484

validation should support all the new configurations:
 https://github.com/conan-io/conan-center-index/issues/25483

Zip files with executables need an extra flag:
 https://github.com/conan-io/conan-center-index/issues/25482

datalogics-robb commented:
We have had this in production for a while using Conan 1.64.1, so we have confidence in the recipe changes here.

valgur commented on recipes/zulu-openjdk/all/conanfile.py:
```diff
@@ -16,30 +16,33 @@ class ZuluOpenJDK(ConanFile):
+        if srcs is None:
+            raise ConanInvalidConfiguration(f"Unsupported os ({self._settings_host.os}). "
+                                            f"This version {self.version} currently does not support"
+                                            f" {self._settings_host.arch} on {self._settings_host.os})")
```
```suggestion
                                            f" {self.settings.arch} on {self.settings.os})")
```
You can use `self.settings` instead (here and elsewhere), which corresponds to host-context settings and is always available.

valgur commented on recipes/zulu-openjdk/all/conanfile.py:
```diff
@@ -13,33 +13,32 @@ class ZuluOpenJDK(ConanFile):
-    @property
-    def _settings_build(self):
-        return getattr(self, "settings_build", self.settings)
+    settings = "os", "arch"
```
```suggestion
    settings = "os", "arch", "compiler", "build_type"
```
Keep the original settings.

datalogics-robb replied:
@valgur can you comment as to why we'd keep the settings here?  My thinking is that we don't use a compiler, so we don't need it's settings.

datalogics-robb replied:
@valgur Just wanted to see if I can bring this back to your attention.
the JDK is a binary package - containing executables and a C library.  For those reasons we wouldn't want the package_id to depend on the compiler and build type.

uilianries reviewed (dismissed):
Hello @datalogics-robb :wave: And sorry for the long delay in reviewing your PR!

First, I want to thank you for opening those issues with detailed information, it not only helped a lot to understand your scenario, but also to simulate your commented errors.


Now, about your changes related to the following issues:


## Fix for the issue #25482

First, I prepared a small recipe that only downloads the JDK for Solaris and unzip:

```python
from conan import ConanFile
from conan.tools.files import get
from conan.tools.layout import basic_layout


class foobarRecipe(ConanFile):
    name = "azul-zulu11-solaris-sparcv9"
    version = "0.1.0"
    package_type = "application"
    settings = "os", "arch"

    def source(self):
        get(self, "https://cdn.azul.com/zulu/bin/zulu11.54.23-ca-jdk11.0.14-solaris_sparcv9.zip",
             sha256="e728006374a6d42003f29d58ac3d76593a9255b66bea9e865dbd2bed2cb371ce")

    def layout(self):
        basic_layout(self, src_folder="src")  
```

Then, I check its extracted result:

```
$ conan source .
conanfile.py (azul-zulu11-solaris-sparcv9/0.1.0): Calling source() in /tmp/foo/src
conanfile.py (azul-zulu11-solaris-sparcv9/0.1.0): Downloading 219.6MB e728006374a6d42003f29d58ac3d76593a9255b66bea9e865dbd2bed2cb371ce
conanfile.py (azul-zulu11-solaris-sparcv9/0.1.0): Downloaded 36.8MB 16% e728006374a6d42003f29d58ac3d76593a9255b66bea9e865dbd2bed2cb371ce
conanfile.py (azul-zulu11-solaris-sparcv9/0.1.0): Downloaded 69.9MB 31% e728006374a6d42003f29d58ac3d76593a9255b66bea9e865dbd2bed2cb371ce
conanfile.py (azul-zulu11-solaris-sparcv9/0.1.0): Downloaded 110.6MB 50% e728006374a6d42003f29d58ac3d76593a9255b66bea9e865dbd2bed2cb371ce
conanfile.py (azul-zulu11-solaris-sparcv9/0.1.0): Downloaded 145.5MB 66% e728006374a6d42003f29d58ac3d76593a9255b66bea9e865dbd2bed2cb371ce
conanfile.py (azul-zulu11-solaris-sparcv9/0.1.0): Downloaded 182.1MB 82% e728006374a6d42003f29d58ac3d76593a9255b66bea9e865dbd2bed2cb371ce
conanfile.py (azul-zulu11-solaris-sparcv9/0.1.0): Uncompressing zulu11.54.23-ca-jdk11.0.14-solaris_sparcv9.zip to .
conanfile.py (azul-zulu11-solaris-sparcv9/0.1.0): Unzipping 373.8MB, this can take a while

$ file src/zulu11.54.23-ca-jdk11.0.14-solaris/bin/java
src/zulu11.54.23-ca-jdk11.0.14-solaris/bin/java: ELF 64-bit MSB executable, SPARC V9, total store ordering, version 1 (SYSV), dynamically linked, interpreter /usr/lib/sparcv9/ld.so.1, not stripped
$ ls -lh src/zulu11.54.23-ca-jdk11.0.14-solaris/bin/java
-rw-rw-r-- 1 uilian uilian 11K Jun 23 08:43 src/zulu11.54.23-ca-jdk11.0.14-solaris/bin/java
```

Just like you commented in the issue, the Java executable has no execution permission.

Now, to cross-check with the regular process, I made:

```
$ wget https://cdn.azul.com/zulu/bin/zulu11.54.23-ca-jdk11.0.14-solaris_sparcv9.zip
--2025-06-23 08:54:03--  https://cdn.azul.com/zulu/bin/zulu11.54.23-ca-jdk11.0.14-solaris_sparcv9.zip
Resolving cdn.azul.com (cdn.azul.com)... 172.64.152.44, 104.18.35.212, 2606:4700:4400::6812:23d4, ...
Connecting to cdn.azul.com (cdn.azul.com)|172.64.152.44|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 219554834 (209M) [application/zip]
Saving to: ‘zulu11.54.23-ca-jdk11.0.14-solaris_sparcv9.zip’

zulu11.54.23-ca-jdk11.0.14-solaris_sparcv9.zip              100%[=========================================================================================================================================>] 209,38M  3,27MB/s    in 61s     

2025-06-23 08:55:05 (3,41 MB/s) - ‘zulu11.54.23-ca-jdk11.0.14-solaris_sparcv9.zip’ saved [219554834/219554834]

$ sha256sum zulu11.54.23-ca-jdk11.0.14-solaris_sparcv9.zip 
e728006374a6d42003f29d58ac3d76593a9255b66bea9e865dbd2bed2cb371ce  zulu11.54.23-ca-jdk11.0.14-solaris_sparcv9.zip
$ unzip -q zulu11.54.23-ca-jdk11.0.14-solaris_sparcv9.zip
$ file zulu11.54.23-ca-jdk11.0.14-solaris/bin/java
zulu11.54.23-ca-jdk11.0.14-solaris/bin/java: ELF 64-bit MSB executable, SPARC V9, total store ordering, version 1 (SYSV), dynamically linked, interpreter /usr/lib/sparcv9/ld.so.1, not stripped
$ ls -l  zulu11.54.23-ca-jdk11.0.14-solaris/bin/java
-rwxr-xr-x 1 uilian uilian 10824 Jan 14  2022 zulu11.54.23-ca-jdk11.0.14-solaris/bin/java*
```

As well pointed, the Conan tool `files.get` is not preserving the original permissions, instead, it is removing them, so using explicit `keep_permission=True` makes sense for this recipe.
Originally, the `tools.files.get()` forwards the `keep_permission` to `tools.files.unzip`: https://docs.conan.io/2/reference/tools/files/basic.html#conan-tools-files-unzip


## Fix for the issue #25483 

Just the previous issue, the current recipe in the ConanCenteIndex does not reflect your personal scenario: the recipe is not covering SunOS nor x86 arch, but only x86_64.

Still, I see it as a positive change, as it keeps the recipe untouched in the future, in case adding new architectures in the conandata.yml, avoiding rebuilding all previous versions available already in the Conan Center.

I can see the Azul repository offers binaries for x86: https://cdn.azul.com/zulu/bin

```
--- 	zulu11.37.19-ca-fx-jre11.0.6-win_i686.zip 	2023-08-17 17:43 	58M
```

So I'm fine with change as well. :+1: 

## Fix for the issue #25484

The issue is plausible, as initially the Zulu JDK package was expected as a tool requirement only, that why it's using `build_settings` only in Conan 2.x., but with an idiom for Conan 1.x
However, with Conan 2.x only and [host and build contexts](https://docs.conan.io/2/tutorial/consuming_packages/cross_building_with_conan.html) used by default, 
the `settings.os` and `settings.arch` should look up according to the profile and requirement scope (requires or tool_requires).

So, when consuming zulu-jdk as a `tool_requirement`, the `settings.os` and `settings.arch` from your build profile will be used.
But in case installing it as a regular requirement for an application or library, it will use your host profile.
This approach is the same as used for the CMake [binaries recipe](https://github.com/conan-io/conan-center-index/blob/master/recipes/cmake/binary/conanfile.py) today in ConanCenterIndex.

----

Thank you again for your contribution and pushing these fixes! Regads!

datalogics-robb commented:
@valgur I'll try to keep an eye on this and remind people to not forget about it.

perseoGI reviewed (dismissed):
Following @uilianries' detailed response, I agree on the benefits this pull request (PR) will bring to the recipe.

Regarding issue [25482](https://github.com/conan-io/conan-center-index/issues/25482), it is clear that maintaining permissions is essential for correctly packaging executables with execution permissions enabled.

For issue [25483](https://github.com/conan-io/conan-center-index/issues/25483), this change will not result in any behavioral changes since all versions except for `11.0.24` only support the `x86_64` and `armv8` architectures. However, it is a significant improvement that will help avoid your issue and automate the validation of supported architectures in the future without needing to update the recipe revision.

As for issue [25484](https://github.com/conan-io/conan-center-index/issues/25484), this is definitely a breaking change for the recipe. I believe this change will ultimately be beneficial, and I also agree that the original recipe was not designed to support this specific use case. This behavior was introduced in Conan v1, trying to support cross-compilation. In conan v2, this has changed as @uilianries pointed out.

jcar87 commented:
Thanks @datalogics-robb  and apologies for the delay in fully reviewing this.

We are partially merging some fixes in this PR: https://github.com/conan-io/conan-center-index/pull/28459 - you are absolutely right that the settings are not being used properly.

For solaris or 32-bit linux: the recipe does not support either of these. I would encourage you to amend the PR to add those in the `conandata.yml`.

For the permissions issue: @uilianries, @perseoGI  - I would first like to understand why this is *NOT* an issue on macOS or Linux - I can see on Linux that the executable flag is correctly preserved.

datalogics-robb committed:
Add a few more JDK versions and platforms.

Version 17.0.15 with x86 versions for Windows and Linux.
Version 21.0.8 on supported platforms, and add Solaris jdk support to 11.0.15.

uilianries commented:
Correct, that's an expected behavior when inflating zip files only: https://github.com/conan-io/conan/blob/2c628aeae4ec3d76d5afbceb2db6439f1af32ad6/conan/tools/files/files.py#L356

We usually add .tar files in CCI and this PR is particular, because downloads a pre-built binary and uses a .zip file instead. Still, we have a few cases in CCI where we use chmod to guarantee the execution permission.
